### PR TITLE
[SymmetricMemory] put socket files in /tmp

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130842
* __->__ #130757
* #130756

Currently the socket files are put in the current directory, which may not be writable in all environments.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/CUDASymmetricMemory.cu [hunk 1]
```diff
@@ -105,7 +109,7 @@ class IpcChannel {
  private:
   static std::string get_socket_name(int pid) {
     std::ostringstream oss;
-    oss << "symm_mem-" << pid;
+    oss << "/tmp/symm_mem-" << pid;
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:112]: Can we respect TMPDIR env variables?

**Review by Chillee — APPROVED** [review]

File: torch/csrc/distributed/c10d/CUDASymmetricMemory.cu [hunk 2]
```diff
@@ -108,8 +112,15 @@ class IpcChannel {
 
  private:
   static std::string get_socket_name(int pid) {
+    const char* tmp_dir = "/tmp";
+    for (const char* env_var : {"TMPDIR", "TMP", "TEMP", "TEMPDIR"}) {
+      if (const char* path = getenv(env_var)) {
+        tmp_dir = path;
+        break;
+      }
+    }
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:121]: You should use std::filesystem::temp_directory_path is what should be used to get this properly in stdlib.

Comment (yifuwang) [reply torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:121]: Merged before I saw this. Addressed in https://github.com/pytorch/pytorch/pull/130842 (also replacing another occurrence).

**Pull request closed without merging**